fs2loader/src/main/scala/com.snowplowanalytics.snowplow.storage.bigquery.fs2loader/Fs2Loader.scala

```diff
@@ -0,0 +1,197 @@
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.storage.bigquery.fs2loader
+
+import cats.syntax.all._
+import cats.effect.{Concurrent, ContextShift, ExitCode, IO, Timer}
+
+import fs2.{Pipe, Stream}
+import fs2.concurrent.Queue
+
+import com.snowplowanalytics.snowplow.badrows.BadRow
+import com.snowplowanalytics.snowplow.storage.bigquery.common.Codecs.toPayload
+import com.snowplowanalytics.snowplow.storage.bigquery.common.Config.Environment
+import com.snowplowanalytics.snowplow.storage.bigquery.fs2loader.events._
+import com.snowplowanalytics.snowplow.storage.bigquery.fs2loader.sinks.PubSub
+import com.snowplowanalytics.snowplow.storage.bigquery.fs2loader.sinks.PubSub.{WriteBadRow, WriteObservedTypes}
+import com.snowplowanalytics.snowplow.storage.bigquery.loader.LoaderRow
+
+import org.slf4j.LoggerFactory
+
+import com.snowplowanalytics.snowplow.analytics.scalasdk.Data.ShreddedType
+
+import scala.concurrent.duration._
+
+object Fs2Loader {
+  private val MaxConcurrency = 5
+  private val GroupByN       = 5
+  private val TimeWindow     = 2.seconds
+
+  /**
+    * Aggregate observed types when a specific number is reached or time passes, whichever happens first.
+    * @param groupByN Number of elements that trigger aggregation.
+    * @param timeWindow Time window to aggregate over if n limit is not reached.
+    * @tparam ST In a prod setting, this should be [[ShreddedType]].
+    * @return A pipe that does not change the type of elements but discards non-unique values within the group.
+    */
+  def aggregateTypes[ST](
+    groupByN: Int,
+    timeWindow: FiniteDuration
+  )(implicit T: Timer[IO], C: Concurrent[IO]): Pipe[IO, Set[ST], Set[ST]] =
+    _.groupWithin(groupByN, timeWindow).map { chunk =>
+      chunk.toList.toSet.flatten
+    }
+
+  /**
+    * Extract observed types from loader rows and add aggregates to a queue.
+    *
+    * @param queue     A queue to add aggregated types to.
+    * @param getTypes A function that extracts types from loader rows.
+    * @param groupByN Number of elements that trigger aggregation.
+    * @param timeWindow Time window to aggregate over if n limit is not reached.
+    * @tparam LR The type of the row, should be [[LoaderRow]] in prod.
+    * @tparam ST The type of the observed types, should be [[ShreddedType]] in prod.
+    * @tparam U The return type of the load function, should be [[Unit]] in prod.
```

Comment:
As we discussed on the meeting, there's no need for generic `U`. It's okay to have them as `Unit` and you still can use queues (or even file sinks) to test it.